text/0000-unnamed-fields.md

```diff
@@ -0,0 +1,463 @@
+name, allowing disambiguation via a single token of context (`union {`).
+
+## Layout and Alignment
+
+The layout and alignment of a `struct` or `union` containing unnamed
+fields should look the same as if each unnamed field has a separately declared
+type and a named field of that type, rather than as if the fields appeared
+directly within the containing `struct` or `union`. In some cases, this may
+result in different alignment.
+
+## Simultaneous Borrows
+
+An unnamed `struct` within a `union` should behave the same with respect to
+borrows as a named and typed `struct` within a `union`, allowing borrows of
+multiple fields from within the `struct`, while not permitting borrows of other
+fields in the `union`.
+
+## Visibility
+
+Each field within an unnamed `struct` or `union` may have an attached
+visibility (`pub` or `pub(crate)`). An unnamed field itself does not have its
+own visibility; all of its fields appear directly within the containing
+structure, and their own visibilities apply.
+
+## Documentation
+
+Public fields within an unnamed `struct` or `union` should appear in the
+rustdoc documentation of the outer structure, along with any doc comment or
+attribute attached to those fields. The rendering should include all unnamed
+fields that contain (at any level of nesting) a public field, and should
+include the `// some fields omitted` note within any `struct` or `union` that
+has non-public fields, including unnamed fields.
+
+Any unnamed field that contains only non-public fields should be omitted
+entirely, rather than included with its fields omitted. Omitting an unnamed
+field should trigger the `// some fields omitted` note.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This introduces additional complexity in structure definitions. Strictly
+speaking, C interfaces do not *require* this mechanism; any such interface
+*could* define named struct or union types, and define named fields of that
+type. This RFC provides a usability improvement for such interfaces.
+
+# Rationale and Alternatives
+[alternatives]: #alternatives
+
+## Not implementing this feature at all
+
+Choosing not to implement this feature would force binding generators (and the
+authors of manual bindings) to invent new names for these groupings of fields.
+Users would need to look up the names for those groupings, and would not be
+able to rely on documentation for the underlying interface. Furthermore,
+binding generators would not have any basis on which to generate a meaningful
+name.
+
+## Not implementable as a macro
+
+We cannot implement this feature as a macro, because it affects the names used
+to reference the fields contained within an unnamed field. A macro could
+extract and define types for the unnamed fields, but that macro would have to
+give a name to those unnamed fields, and accesses would have to include the
+intermediate name.
+
+## Unnamed fields using `_` with named types
```

Comment:
@RalfJung 
>we're not even talking about copying a proper standard, but a proprietary vendor extension

For the record, this feature of C is officially standardized as of C2011: see section 6.7.2.1, paragraphs 13 and 19, of [document N1570](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1570.pdf). (Document N1570 is the closest approximation to the 2011 revision of the C standard that is publicly available at no charge.)  It might be easier to understand the feature as-standardized from its proposal, which is [document N1406](http://www.open-std.org/jtc1/sc22/wg14/www/docs/).  According to N1406, it is also a standard C++ feature, but I am not nearly as familiar with either the C++ standard or its committee's process, so I cannot point you at specifics there.

> clearly large amounts of C code don't need this

To my mind, the strongest argument for including this feature in Rust is that both Windows and POSIX use this feature to define types that are part of their public, standardized API.  This means "large amounts of C" are in fact using the feature without this being evident from their own code. Several examples have already appeared upthread, but I'll reiterate the issue with `siginfo_t`, since it's the one I understand best: the documented type signature of this struct is

``` c
typedef struct {
    int si_signo;
    int si_code;
    int si_errno;
    pid_t si_pid;
    uid_t si_uid;
    void *si_addr;
    int si_status;
    long si_band;
    union sigval si_value;
} siginfo_t;
```

but, on three out of four of the different Unixes I can conveniently check (Linux, NetBSD, Solaris; not OSX) some subset of those fields are actually provided by an anonymous union. And it's a different subset with a different "true" layout on each.  Therefore, any portable C program that uses `siginfo_t` has no choice but to use this feature -- but the programmer is not aware of this.  [codesearch.debian.net](https://codesearch.debian.net/search?q=filetype%3Ac+%5C.si_%28signo%7Ccode%7Cerrno%7Cpid%7Cuid%7Caddr%7Cstatus%7Cband%7Cvalue%29%5Cb+-package%3Aglibc+-package%3Agcc-5+-package%3Agcc-6+-package%3Agcc-7+-package%3Ahurd+-package%3Alinux+-package%3Amusl+-package%3Auclibc+-package%3Adietlibc+-package%3Agnulib+-package%3Alibnih) finds 3200 uses of these fields in 98 packages, after excluding the packages (e.g. "linux" and "glibc") that provide its definition.

(Footnote: some C libraries don't actually use anonymous unions here, instead they use named unions and preprocessor macros that expand to a string of field accessors. I hope you will agree that this is strictly worse, since C preprocessor macros have no scope control.  I would actually endorse "improve Rust's macro system to permit field names to be macros, on a per-type basis" as an alternative to this proposal, provided this could be invisible (no `!` sigil).)